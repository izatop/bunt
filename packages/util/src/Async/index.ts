export * from "./fn";
export * from "./AsyncState";
export * from "./AsyncCallback";
export * from "./Defer";
export * from "./AsyncPushPull";
