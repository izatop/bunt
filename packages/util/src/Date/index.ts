export * from "./DateTime.js";
export * from "./Locale.js";
export * from "./interface.js";
