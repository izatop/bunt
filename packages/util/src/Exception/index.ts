export * from "./AssertionError";
export * from "./interfaces";
