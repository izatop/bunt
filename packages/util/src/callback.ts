export function resolveOrReject(resolve: (value: void | PromiseLike<void>) => void, reject: (reason?: any) => void) {
    return (error?: Error): void => {
        if (error) {
            reject(error);
            
            return;
        }

        resolve();
    };
}

export function filterValueCallback<T>(test: (value: unknown) => boolean, cb: (value: T) => unknown) {
    return (value: T): void => {
        if (test(value)) {
            cb(value);
        }
    };
}
