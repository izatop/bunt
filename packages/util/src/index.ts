export * from "./is";
export * from "./stream";
export * from "./assert";
export * from "./object";
export * from "./string";
export * from "./decorator";
export * from "./function";
export * from "./callback";
export * from "./Date";
export * from "./Transform";
export * from "./Logger";
export * from "./Exception";
export * from "./Node";
export * from "./Perf";
export * from "./Async";
export * from "./Ref";
export * from "./Map";
export * from "./interfaces";
