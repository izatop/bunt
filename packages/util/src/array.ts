import {all} from "./Async";
import {Promisify} from "./interfaces";

export async function safeMap<T>(values: Iterable<T>, fn: (value: T) => Promisify<any>): Promise<unknown[]> {
    const result: Promise<unknown>[] = [];
    for (const value of values) {
        try {
            result.push(await fn(value));
        } catch (error) {
            result.push(Promise.resolve(undefined));
        }
    }

    return all(result);
}
