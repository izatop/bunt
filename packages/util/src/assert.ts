import {AssertionError} from "./Exception";
import {isFunction} from "./is";

export type AssertionDetailsAllowType = string | object | null | number;
export type AssertionDetails = (() => AssertionDetailsAllowType) | AssertionDetailsAllowType;
export type AssertionMessage = string | (() => string);

export function assert(expr: any, message?: AssertionMessage, details?: AssertionDetails): asserts expr {
    if (!expr) {
        throw new AssertionError(isFunction(message) ? message() : message, details);
    }
}

export function fails(expr: any, message?: AssertionMessage, details?: AssertionDetails) {
    if (!!expr) {
        throw new AssertionError(isFunction(message) ? message() : message, details);
    }
}

export function pass<T>(value: any, transform: (v: any) => T, validator: (v: any) => boolean, label?: string): T {
    assert(value, `${label || "Value"} should be defined`);

    const nextValue = transform(value);
    assert(
        validator(nextValue),
        `This ${label || "value"} "${nextValue}" should pass the validation stage: ${validator.toString()}`,
    );

    return nextValue;
}
