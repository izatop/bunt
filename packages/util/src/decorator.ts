import {assert} from "@bunt/assert";
import {isFunction, isUndefined} from "@bunt/is";
import {DecoratorTarget} from "@bunt/type";

export function memoize(target: DecoratorTarget, key: string): PropertyDescriptor {
    const descriptor = Reflect.getOwnPropertyDescriptor(target, key);
    assert(!isUndefined(descriptor));
    assert(
        isFunction(descriptor.value) || isFunction(descriptor.get),
        "Memoization should be apply only to methods",
    );

    const memoizeCache = new WeakMap();
    if (descriptor.value) {
        const fn = descriptor.value;
        descriptor.value = function memoizeCacheFunction(this: any, ...args: any[]): any {
            if (!memoizeCache.has(this)) {
                memoizeCache.set(this, fn.call(this, args));
            }

            return memoizeCache.get(this);
        };
    }

    if (descriptor.get) {
        const fn = descriptor.get;
        descriptor.get = function memoizeCacheFunction(this: any): any {
            if (!memoizeCache.has(this)) {
                memoizeCache.set(this, fn.call(this));
            }

            return memoizeCache.get(this);
        };
    }

    return descriptor;
}

export function bind<T extends(...args: any[]) => any>(
    target: Record<any, any>,
    propertyKey: string,
    descriptor: TypedPropertyDescriptor<T>): TypedPropertyDescriptor<T> | void {
    return {
        configurable: true,
        get(this: T): T {
            const bound = descriptor.value?.bind(this) as T;
            if (bound) {
                Object.defineProperty(this, propertyKey, {
                    value: bound,
                    configurable: true,
                    writable: true,
                });
            }

            return bound;
        },
    };
}
