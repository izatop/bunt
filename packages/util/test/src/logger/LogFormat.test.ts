import {debugLogFormat, defaultLogFormat, readableJSONLogFormat, SeverityLevel} from "../../../src";

describe("Log format", () => {
    test("test default log formatter", () => {
        expect(defaultLogFormat({
            pid: 5761,
            label: "Test",
            message: "Log message",
            args: [1, {foo: "123"}],
            groupId: "groupId",
            host: "srv1",
            timestamp: 1585574245816,
            severity: SeverityLevel.EMERGENCY,
        })).toMatchSnapshot();
    });

    test("test readable JSON log formatter", () => {
        expect(readableJSONLogFormat({
            pid: 5761,
            label: "Test",
            message: "Log message",
            args: [1, {foo: "123"}],
            groupId: "groupId",
            host: "srv1",
            timestamp: 1585574245816,
            severity: SeverityLevel.NOTICE,
        })).toMatchSnapshot();
    });

    test("test debug log formatter", () => {
        expect(debugLogFormat({
            pid: 6611,
            label: "Test",
            message: "Log message",
            args: [1, {foo: "123"}],
            groupId: "G-7782",
            host: "srv2",
            timestamp: 1585574245816,
            severity: SeverityLevel.INFO,
        })).toMatchSnapshot();
    });
});
