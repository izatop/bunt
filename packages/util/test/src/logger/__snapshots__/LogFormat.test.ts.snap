// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Log format test debug log formatter 1`] = `
"2020-03-30T13:17:25.816Z INFO host=srv2 pid=6611 label=Test group=G-7782
Log message
[
  1,
  {
    \\"foo\\": \\"123\\"
  }
]
"
`;

exports[`Log format test default log formatter 1`] = `"{\\"pid\\":5761,\\"label\\":\\"Test\\",\\"message\\":\\"Log message\\",\\"args\\":[1,{\\"foo\\":\\"123\\"}],\\"groupId\\":\\"groupId\\",\\"host\\":\\"srv1\\",\\"timestamp\\":1585574245816,\\"severity\\":0}"`;

exports[`Log format test readable JSON log formatter 1`] = `
"{
  \\"pid\\": 5761,
  \\"label\\": \\"Test\\",
  \\"message\\": \\"Log message\\",
  \\"args\\": [
    1,
    {
      \\"foo\\": \\"123\\"
    }
  ],
  \\"groupId\\": \\"groupId\\",
  \\"host\\": \\"srv1\\",
  \\"timestamp\\": 1585574245816,
  \\"severity\\": 5
}"
`;
